src/core/qgspointlocator.h

```diff
@@ -153,12 +153,13 @@ class CORE_EXPORT QgsPointLocator : public QObject
     enum Type
     {
       Invalid = 0, //!< Invalid
-      Vertex  = 1, //!< Snapped to a vertex. Can be a vertex of the geometry or an intersection.
-      Edge    = 2, //!< Snapped to an edge
-      Area    = 4, //!< Snapped to an area
-      Centroid = 8, //!< Snapped to a centroid
-      MiddleOfSegment = 16, //!< Snapped to the middle of a segment
-      All = Vertex | Edge | Area | Centroid | MiddleOfSegment //!< Combination of all types
+      Vertex  = 1 << 0, //!< Snapped to a vertex. Can be a vertex of the geometry or an intersection.
+      Edge    = 1 << 1, //!< Snapped to an edge
+      Area    = 1 << 2, //!< Snapped to an area
+      Centroid = 1 << 3, //!< Snapped to a centroid
+      MiddleOfSegment = 1 << 4, //!< Snapped to the middle of a segment
+      LineEndpoint = 1 << 5, //!< Start or end points of lines only (since QGIS 3.20)
+      All = Vertex | Edge | Area | Centroid | MiddleOfSegment //!< Combination of all types. Note LineEndpoint is not included as endpoints made redundant by the presence of the Vertex flag.
```

Comment:
*LineEndpoint * is not redundant for polygon layers, since *Vertex* will not snap on ring first/last vertex!

Reply:
I don't understand -- if you're snapping to all vertices then that natural includes all ring first/last vertices..

Reply:
Haven't woken completely yet, forget this ever happened!! :laughing: